sdk/go/docker/docker.go

```diff
@@ -0,0 +1,548 @@
+
+	stdout, err := runCommandThatCanFail(cmd, args, logResource, reportFullCommandLine, false, stdin, env)
+
+	if err != nil {
+		return "", errors.Wrapf(err, "%s failed with error",
+			getCommandLineMessage(cmd, args, reportFullCommandLine, env))
+	}
+
+	return stdout, nil
+}
+
+func runBasicCommandThatMustSucceed(cmd string, args []string, logResource pulumi.Resource) (string, error) {
+	return runCommandThatMustSucceed(cmd, args, logResource, true, "", nil)
+}
+
+// Runs a CLI command in a child process, returning a promise for the process's exit. Both stdout
+// and stderr are redirected to process.stdout and process.stder by default.
+//
+// If the [stdin] argument is defined, it's contents are piped into stdin for the child process.
+//
+// [logResource] is used to specify the resource to associate command output with. Stderr messages
+// are always sent (since they may contain important information about something that's gone wrong).
+// Stdout messages will be logged ephemerally to this resource.  This lets the user know there is
+// progress, without having that dumped on them at the end.  If an error occurs though, the stdout
+// content will be printed.
+//
+// The promise returned by this function should never reach the rejected state.  Even if the
+// underlying spawned command has a problem, this will result in a resolved promise with the
+// [CommandResult.code] value set to a non-zero value.
+func runCommandThatCanFail(cmdName string, args []string, logResource pulumi.Resource, reportFullCommandLine bool,
+	reportErrorAsWarning bool, stdinInput string, env map[string]string) (string, error) {
+
+	logging.InitLogging(true, 7, false)
+
+	// Let the user ephemerally know the command we're going to execute.
+	logging.Infof("Executing " + getCommandLineMessage(cmdName, args, reportFullCommandLine, env))
+
+	// Generate a unique stream-ID that we'll associate all the docker output with. This will allow
+	// each spawned CLI command's output to associated with 'resource' and also streamed to the UI
+	// in pieces so that it can be displayed live.  The stream-ID is so that the UI knows these
+	// messages are all related and should be considered as one large message (just one that was
+	// sent over in chunks).
+	//
+	// We use rand here in case our package is loaded multiple times in memory (i.e. because
+	// different downstream dependencies depend on different versions of us).  By being random we
+	// effectively make it completely unlikely that any two cli outputs could map to the same stream
+	// id.
+	//
+	// Pick a reasonably distributed number between 0 and 2^30.  This will fit as an int32
+	// which the grpc layer needs.
+	streamID := rand.Int()
+
+	cmd := exec.Command(cmdName, args...)
+
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		logging.Errorf("Error retrieving stderr: %v", err)
+		return "", err
+	}
+
+	stdout, err := cmd.StdoutPipe()
```

Comment:
do we need to close stdout and stderr  like we do with the stdinPipe below?

Reply:
`StderrPipe` and `StdoutPipe` close automatically with `cmd.Wait`. If I understand correctly, `StdinPipe` will also close with `cmd.Wait`, but we close it manually because there's no reason to leave it open until `cmd.Wait` exits. See https://golang.org/pkg/os/exec/#Cmd.StderrPipe.